### DISCUSSION TITLE
Connection timeout and recurring queries

### DISCUSSION THREAD

**Discussion by sehmaschine** [prisma/prisma#5384] [Q&A]
We're on DigitalOcean using a managed database and run into connection issues. One background tasks fails (due to a timeout), but we still get lots of connections and queries after the tasks has finished (although we disconnect prisma).

I have no idea what happens, but here's a simplified version of the code we are running with our job:

```js
const { PrismaClient } = require("@prisma/client");
const prisma = new PrismaClient();

async function start() {
  const files = await prisma.data_reportfile.findMany({
      take: 10,
      where: {
        OR: [
          {
            data_report: { date: { equals: null }},
          },
          {
            data_report: { data_report_countries: { none: { id: undefined }}},
          }
        ],
        AND: {
          analysis_sent: false,
        },
      },
      select: {
        id: true,
        file_url: true,
        data_report: {
          select: {
            id: true,
            hash: true,
          }
        }
      }
    });
  console.log(files.length);
}

(async () => {
  start()
    .catch((e) => console.error(e))
    .finally(async () => {
      await prisma.$disconnect();
      process.exit(0);
    });
})();
```

After running this script, we immediately get up to 7 active connections with our DB and the query keeps on running for hours (until the server shuts down).

Any idea?

**Comment by sehmaschine** [comment]
Funny thing is, I don´t see any prisma process running with our droplet. But we still get lots of queries with the DB. Each query is a bit slower than the previous one. This goes on for hours. Then suddenly all connections are gone. Is there a way to force a disconnect? Or any other way to remove those queries?

**Reply by ryands17** [reply]
@sehmaschine Is the above task being run in the background? If so, could you try by adding a `connection_limit=1` parameter to the connection string of your database URL as mentioned [here](https://www.prisma.io/docs/concepts/database-connectors/postgresql/#connection-url)? This should limit the connections opened by Prisma and as you're already disconnecting, it shouldn't be an issue.

Also the last `async` anonymous function isn't required so that can be omitted as well. Just the `start` method with `catch` and `finally` should do.

**Reply by sehmaschine** [reply]
@ryands17 Yes, it's a background task.

I could add the connection_limit, but that would also affect the non-background process (our main app), right?. Or how can I define multiple connection strings with Prisma?

**Reply by ryands17** [reply]
You can do that in the snippet above by adding the URL in this format:

```ts
const prisma = new PrismaClient({
    datasources: { db: { `${url}?connection_limit=1` } },
})
```

**Reply by sehmaschine** [reply]
I see. Will try that. 

Any idea why we get a timeout in the first place? The query runs locally without any issues (using the exact same DB).

**Reply by ryands17** [reply]
I'm not exactly sure, but is your database and application in the same region? Usually this is one cause for a timeout. I would suggest adding with the `connect_timeout` to a higher number so that it can connect properly.

**Reply by sehmaschine** [reply]
Now we get this:
Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter.

I've already tried to set `connect_timeout` to 10, then 20. Nothing changes.
And we still see with the DB logs that the disconnect doesn't seem to work.

**Reply by sehmaschine** [reply]
It works fine if I simplify the query to something like this:

```js
  const files = await prisma.data_reportfile.findMany({
      take: 10,
      select: {
        id: true,
      }
    });
```

**Reply by ryands17** [reply]
Do you currently just have a single call in the background task or multiple calls running simultaneously as a single call shouldn't have much impact on this that connecting is not possible. Also could you add `DEBUG=*` as your environment variable so that you can actually view if Prisma is being disconnected. You should have something like this in the logs:

![image](https://user-images.githubusercontent.com/19697099/106468653-39249b00-64c4-11eb-8a4d-8140b484b873.png)

If this still doesn't work, I would suggest opening an issue [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) so that we can look into this.

**Reply by sehmaschine** [reply]
Connections are ok now (we're using pgbouncer). It's a single call btw.

But I can only use the simple query. If I change it to the more complex one, the server goes down after about 10 minutes. Really hard to debug since it does not happen locally (and I can't test with the production server since we need it running). But the complex query takes quite a while and that leads to accumulated connections which takes the server down eventually. Not sure if it makes sense to open an issue with this half-baked insight.

**Reply by janpio** [reply]
> But the complex query takes quite a while and that leads to accumulated connections which takes the server down eventually.

Can you elaborate on this a bit? Is the server stalling for other connections as well?

**Reply by sehmaschine** [reply]
No, it´s the only connection we have (right now). What I see with the logs is that once the complex query is triggered, the db connections (as shown with the digitalocean "insights") are going up pretty fast (until they reach 20, which is the pgbouncer limit). But in that case, no further connections (or requests) are possible. The server is basically idle for about an hour. Then the connections slowly decrease and we can add connections again.

Sorry for not being able to add more details. If there's anything I can do to help, please let me know.

**Reply by janpio** [reply]
These cases usually are hard to understand, so we will probably play the question + answer game for some more time :D

I know you have a production database with PgBouncer in front of it and a background task. 
What other queries are being run on this database? With Prisma? By the same app?
How is that background task triggered? Is is possible it is triggered multiple times while the first instance is still running?

If you want we can also chat via slack.prisma.io, I am @janpio there as well.

**Reply by janpio** [reply]
> Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter.

To get around this error, you could add `pool_timeout=0` to your connection string. It will cause queries to wait indefinitely to get a connection from the connection pool - at the cost of queries waiting for a long time when there is no connection. Usually there is a default timeout of 5s that seems to have been triggered for you.

**Reply by sehmaschine** [reply]
> What other queries are being run on this database? With Prisma? By the same app?

None.

> How is that background task triggered? Is is possible it is triggered multiple times while the first instance is still running?

With the test scenario (and the compley query) which lead to the connection issues I triggered the background task manually. So it was only triggered once. With the simple query it is now triggered once per minute automatically.

> To get around this error, you could add pool_timeout=0 to your connection string.

What I did before we used pgbouncer is that I set `connect_timeout` to 20, but that didn't make a difference.

Just an idea, referring to this:
https://www.prisma.io/docs/concepts/components/prisma-client/connection-management
Maybe I should have used `prisma.$connect()` instead of relying on the lazy connection. I'm not exactly sure what happens under the hood, but since the initial/complex query took so long it might have affected the connection.

**Reply by janpio** [reply]
> > To get around this error, you could add pool_timeout=0 to your connection string.
> 
> What I did before we used pgbouncer is that I set `connect_timeout` to 20, but that didn't make a difference.

`connect_timeout` is the at another step in the process (when the connection pool tries to create a connection to the database). The error message you posted above, is from the connection pool and would only be influenced by `pool_timeout`.



> > What other queries are being run on this database? With Prisma? By the same app?
> 
> None.
> 
> > How is that background task triggered? Is is possible it is triggered multiple times while the first instance is still running?
> 
> With the test scenario (and the compley query) which lead to the connection issues I triggered the background task manually. So it was only triggered once. With the simple query it is now triggered once per minute automatically.
> 
> > To get around this error, you could add pool_timeout=0 to your connection string.
> 
> What I did before we used pgbouncer is that I set `connect_timeout` to 20, but that didn't make a difference.

That is super weird.

So you are basically observing that you run 1 (!) Prisma Client query manually, and you are seeing an increasing number of database connections caused by that, correct? And it only happens when you run that query on the production database, not on a local one?


> Just an idea, referring to this:
> [prisma.io/docs/concepts/components/prisma-client/connection-management](https://www.prisma.io/docs/concepts/components/prisma-client/connection-management)
> Maybe I should have used `prisma.$connect()` instead of relying on the lazy connection. I'm not exactly sure what happens under the hood, but since the initial/complex query took so long it might have affected the connection.

This _should_ not make a difference.

**Reply by sehmaschine** [reply]
> So you are basically observing that you run 1 (!) Prisma Client query manually, and you are seeing an increasing number of database connections caused by that, correct? And it only happens when you run that query on the production database, not on a local one?

Exactly. 

We need to run the current tasks for another day, I guess. Then I'm able to do more debugging with the production server and giving `pool_timeout` a try.

**Reply by sehmaschine** [reply]
I added the `pool_timeout` but it doesn't make a difference. Still getting "Headers Timeout Error" and the connections are adding up until the database server eventually crashes.

Not sure how to solve this, but I guess I have to give up at this point. Every time I send this query it takes about 3 hours until the database server is back to normal. And we just cannot let that happen on a regular basis.

Thanks for all your help though!

**Reply by janpio** [reply]
> Still getting "Headers Timeout Error" and

Can you share the exact error?

if you still want to debug further:
Can you share the schema required to run this query and possibly describe the data? Or even share some data privately with us?
If you turn on query logging in Prisma, can you share the generated query with us?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
